File: crypto/src/main/java/io/oversec/one/crypto/encoding/AbstractXCoder.java
package io.oversec.one.crypto.encoding;


import android.content.Context;
import io.oversec.one.crypto.encoding.pad.AbstractPadder;
import io.oversec.one.crypto.proto.Outer;

import java.io.IOException;

public abstract class AbstractXCoder implements IXCoder {


    protected final Context mCtx;

    protected abstract String encodeInternal(Outer.Msg msg, AbstractPadder padder, String packagename) throws IOException;


    // public abstract String getEncodedTextOnly(String encText) throws IOException;

    public AbstractXCoder(Context ctx) {
        mCtx = ctx;
    }

    @Override
    public String encode(Outer.Msg msg, AbstractPadder padder, String plainTextForWidthCalculation, boolean appendNewLines, String packagename) throws IOException {
        if (padder!=null) {
            padder.reset();
        }
        int nn = appendNewLines ? countNewLines(plainTextForWidthCalculation) : 0;
        String internal = encodeInternal(msg,padder,packagename);
        StringBuffer r = new StringBuffer(internal);
        //prependPrefix(r);
        if (nn > 0) {
            appendNewLines(r, nn);
        }
        if (plainTextForWidthCalculation != null && padder != null) {
            padder.pad(plainTextForWidthCalculation, r);
        }
        return r.toString();
    }


    private int countNewLines(String s) {
        int r = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '\n') {
                r++;
            }
        }
        return r;
    }

    private void appendNewLines(StringBuffer sb, int n) {
        for (int i = 0; i < n; i++) {
            sb.append('\n');
        }
    }


}

File: crypto/src/main/java/io/oversec/one/crypto/encoding/AbstractXCoder.kt
package io.oversec.one.crypto.encoding

import android.content.Context
import io.oversec.one.crypto.encoding.pad.AbstractPadder
import io.oversec.one.crypto.proto.Outer
import java.io.IOException

abstract class AbstractXCoder (protected val mCtx: Context) : IXCoder {

    @Throws(IOException::class)
    abstract fun encodeInternal(
        msg: Outer.Msg,
        padder: AbstractPadder?,
        packagename: String
    ): String

    @Throws(IOException::class)
    override fun encode(
        msg: Outer.Msg,
        padder: AbstractPadder?,
        plainTextForWidthCalculation: String?,
        appendNewLines: Boolean,
        packagename: String
    ): String {
        padder?.reset()
        val nn = if (appendNewLines && plainTextForWidthCalculation!=null) (plainTextForWidthCalculation.lines().size-1) else 0
        val internal = encodeInternal(msg, padder, packagename)
        val r = StringBuffer(internal)
        //prependPrefix(r);
        if (nn > 0) {
            r.append("\n".repeat(nn))
        }
        plainTextForWidthCalculation?.let {  padder?.pad(it, r)  }
        return r.toString()
    }


    private fun countNewLines(s: String): Int {
        return s.lines().size-1
    }



}
